File: 300. 最长递增子序列/Solution.go
package main

func main() {

}

func lengthOfLIS(nums []int) int {
	size := len(nums)
	if size == 0 {
		return 0
	}

	res := 1
	ans := make([]int, size)
	ans[0] = 1

	for i := 1; i < size; i++ {
		ans[i] = 1
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] && ans[j]+1 > ans[i] {
				ans[i] = ans[j] + 1
			}
		}
		if ans[i] > res {
			res = ans[i]
		}
	}
	return res
}

/*
	使用动态规划，最重要的是找到状态转移方程
	把之前计算的结果记录下来，简便了后面的计算，最终的到答案
*/

File: 300. 最长递增子序列/Solution.java
class Solution {
    public static void main(String[] args) {
        lengthOfLIS(new int[]{4,10,4,3,8,9});
    }
    public  static int lengthOfLIS(int[] nums) {
        int size = nums.length;
        if (size == 0 || size == 1){
            return size;
        }

        int[] base = new int[size+1];

        base[1] = nums[0];
        int len = 1;

        for (int i = 1; i < size; ++i) {
            int target = nums[i];
            if (target > base[len]){
                len++;
                base[len] = target;
            }else {
                int l = 1, r = len;
                int ans = len;

                while (l <= r){
                    int mid = (l + r) >> 1;
                    if (base[mid] >= target){
                        ans = mid;
                        r = mid - 1;
                    }else {
                        l = mid + 1;
                    }
                }

                base[ans] = target;
            }
        }

        return len;
    }
}